model/map/src/main/java/org/keycloak/models/map/singleUseObject/MapSingleUseObjectProvider.java

```diff
@@ -0,0 +1,213 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.singleUseObject;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.models.ActionTokenKeyModel;
+import org.keycloak.models.ActionTokenStoreProvider;
+import org.keycloak.models.ActionTokenValueModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ModelDuplicateException;
+import org.keycloak.models.SingleUseObjectProvider;
+import org.keycloak.models.map.common.TimeAdapter;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public class MapSingleUseObjectProvider implements ActionTokenStoreProvider, SingleUseObjectProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapSingleUseObjectProvider.class);
+    private final KeycloakSession session;
+    protected final MapKeycloakTransaction<MapSingleUseObjectEntity, ActionTokenValueModel> actionTokenStoreTx;
+
+    public MapSingleUseObjectProvider(KeycloakSession session, MapStorage<MapSingleUseObjectEntity, ActionTokenValueModel> storage) {
+        this.session = session;
+        actionTokenStoreTx = storage.createTransaction(session);
+
+        session.getTransactionManager().enlistAfterCompletion(actionTokenStoreTx);
+    }
+
+    private Function<MapSingleUseObjectEntity, ActionTokenValueModel> singleUseEntityToAdapterFunc() {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return origEntity -> {
+            if (Time.currentTime() < origEntity.getExpiration())  {
+                return new MapSingleUseObjectObjectAdapter(session, origEntity);
+            } else {
+                actionTokenStoreTx.delete(origEntity.getId());
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public void put(ActionTokenKeyModel actionTokenKey, Map<String, String> notes) {
+        if (actionTokenKey == null || actionTokenKey.getUserId() == null || actionTokenKey.getActionId() == null) {
+            return;
+        }
+
+        String sKey = actionTokenKey.getUserId() + ":" + actionTokenKey.getActionId() + ":" + actionTokenKey.getActionVerificationNonce();
```

Comment:
Ideally we'd like to have UUIDs as IDs across all stores. Would it be possible to use different ID than this one? Maybe we can add `USER_ID`, ACTION_ID and ACTION_VERIFICATION_NONCE searchable fields and use those to check whether action token already exist?

Reply:
@vramik Initially, I used those searchable fields but then we agreed on using a composite key and access an entity directly by `id` for performance reasons. Currently, I see hardcoded  `UUID` key type in `JpaMapKeycloakTransaction` so it would fail. Why we can't use configurable `keyConverter` as in `ConcurrentHashMapKeycloakTransaction` and use `String` primary keys for some areas?

Reply:
@vramik The reason for having it this way is performance. For example, HotRod is faster searching by ID than searching by Ickle query.

@martin-kanis Another solution would be to make UUID out of the three fields (`UUID.nameUUIDFromBytes`). It makes MD5 hash in the background. However, we need to count on the fact that we can have duplicities with this approach, therefore we need to be able to store more entries for one UUID and filter the correct one in Java.

On the other hand, this ^ can cause performance problems for JPA (for JPA it could be faster to make SQL with all three keys as we can have a compound index on them). This is getting a little bit more complicated, maybe we could have an MCB `read` call on the level of the provider, and then translate the criteria to `readById` for HotRod and CHM, but use as is for JPA. WDYT?

Reply:
@martin-kanis, @mhajas thank you!

> Why we can't use configurable `keyConverter` as in `ConcurrentHashMapKeycloakTransaction` and use `String` primary keys for some areas?

We can do that. Until now there has been no reason for it. The implications would have to be investigated though.

Another possible solution could be to use `@EmbeddedId` for this specific area. Again this option would need to be investigated further.

> maybe we could have an MCB read call on the level of the provider, and then translate the criteria to readById for HotRod and CHM, but use as is for JPA. WDYT?

I like the idea, if it would be possible then it'd the easiest way for JPA implementation.